### PR TITLE
librbd: s/boost::variant/std::variant/

### PR DESCRIPTION
**Pull request by tchaikov**
boost::variant explicitly prevent it from being compared other types
by marking the return type of, for instance operator==(const T&) as
"void", if T is not identical to the variant type. so address this,
let's use std::variant<> instead.

more standard compliant and simpler this way.

because of https://cplusplus.github.io/LWG/issue3052, libstdc++
only specializes for std::variant<> for std::visit(), while libc++
allows the derived types of std::variant<> to be "visited", hence
we add an adaptor for SnapshotNamespace.

Signed-off-by: Kefu Chai <tchaikov@gmail.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
this change is part of the bigger initiative of #45133. i am extracting it out and proposing it separately because it is an improvement even if we were not looking at moving to C++20, also hope that it can be reviewed and merged sooner.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test api

**Review by cbodley — COMMENTED** [review]
lgtm. will defer to @idryomov for approval

**Comment by tchaikov** [comment]
```
2022-03-03 15:23:53,927.927 INFO:__main__:Creating cluster with 1 MDS daemons
2022-03-03 15:23:53,928.928 INFO:__main__:
tearing down the cluster...
2022-03-03 15:23:55,603.603 INFO:__main__:
ceph cluster torn down
2022-03-03 15:23:55,612.612 INFO:__main__:
running vstart.sh now...
Build step 'Execute shell' marked build as failure
```

**Comment by tchaikov** [comment]
@idryomov hi Ilya, could you help review this change? Just wanted to migrate our source tree to C++20 piecemeal.

File: src/cls/rbd/cls_rbd_types.h [hunk 1]
```diff
@@ -674,20 +693,20 @@ struct SnapshotNamespace : public SnapshotNamespaceVariant {
   void decode(ceph::buffer::list::const_iterator& it);
```
Comment (trociny) [inline src/cls/rbd/cls_rbd_types.h:693]: (this comment is for the line 669, `SnapshotNamespace(T&& t` constructor)

So, as I understand your "test/rbd_mirror: use ASSERT_TRUE() to compare iterators" commit log message, this constructor does not work any more.

Is there any way to actually make it work? Then it seems we would not need changing the tests to use `ASSERT_TRUE` (I don't mind of using `ASSERT_TRUE`, but I am afraid someone, may be me, will hit this again when adding new tests and will have to spend time figuring out why it does not work).

And if it is not possible to make the constructor work, then may be it is better to remove it? Or may be I completely misunderstood the issue?

Comment (tchaikov) [reply src/cls/rbd/cls_rbd_types.h:688]: @trociny hi Mykola, thanks for your review.

the reason why i wanted to replace `boost::variant<>` with `std::variant<>` was, as the commit message explains, `boost::variant` does not allow us to compare a `boost::variant` with its derived class. i think the purpose of `SnapshotNamespace(T&& t)` is to convert a type in the "union" to `SnapshotNamespace`. for instance, this template constructor allows us to construct a `SnapshotNamespace` from `UserSnapshotNamespace`.

the `ASSERT_TRUE()` fix is all about the printability of iterators. it seems that the templated constructor looks like a honey pot from the compilers perspective, and it just insists on using it when looking up the candidates of `operator<<(..)`. that's why it fails to fallback to the fallback printer provided by GTest.

anyhow, i changed the way how the constructor is implemented. seems it's a better solution

- no need to have the `ASSERT_TRUE()` change
- the constructor is also simpler.

File: src/cls/rbd/cls_rbd_types.h [hunk 2]
```diff
@@ -656,7 +675,7 @@ std::ostream& operator<<(std::ostream& os, const TrashSnapshotNamespace& ns);
 std::ostream& operator<<(std::ostream& os, const MirrorSnapshotNamespace& ns);
 std::ostream& operator<<(std::ostream& os, const UnknownSnapshotNamespace& ns);
 
-typedef boost::variant<UserSnapshotNamespace,
+typedef std::variant<UserSnapshotNamespace,
                        GroupSnapshotNamespace,
                        TrashSnapshotNamespace,
                        MirrorSnapshotNamespace,
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/cls/rbd/cls_rbd_types.h]: sure, reformatted.

**Review by trociny — APPROVED** [review]
LGTM

File: src/cls/rbd/cls_rbd_types.cc [hunk 3]
```diff
@@ -1124,6 +1129,16 @@ void TrashImageSpec::dump(Formatter *f) const {
...
 
+std::ostream& operator<<(std::ostream& os, const TrashImageSpec& spec)
+{
+  return os << "["
+	    << "source=" << spec.source << ", "
+	    << "name=" << spec.name << ", "
+	    << "deletion_time=" << spec.deletion_time << ", "
+	    << "deferment_end_time=" << spec.deferment_end_time
+	    << "]";
+}
```
Comment (idryomov) [inline src/cls/rbd/cls_rbd_types.cc:1140]: This seems unrelated, why did it become needed?

**Review by idryomov — CHANGES_REQUESTED** [review]

**Comment by idryomov** [comment]
> because of https://cplusplus.github.io/LWG/issue3052, libstdc++
> only specializes for std::variant<> for std::visit(), while libc++
> allows the derived types of std::variant<> to be "visited", hence
> we add an adaptor for SnapshotNamespace.

So libstdc++ isn't compliant here?  Because https://en.cppreference.com/w/cpp/utility/variant/visit says "... or a (possibly const-qualified) class C such that there is exactly one [std::variant](https://en.cppreference.com/w/cpp/utility/variant) specialization that is a base class of C and it is a public and unambiguous base class."

File: src/librbd/mirror/snapshot/Utils.cc [hunk 4]
```diff
@@ -28,7 +28,7 @@ bool get_rollback_snap_id(
     uint64_t *rollback_snap_id) {
 
   for (; it != end; it++) {
-    auto mirror_ns = boost::get<cls::rbd::MirrorSnapshotNamespace>(
+    auto mirror_ns = std::get_if<cls::rbd::MirrorSnapshotNamespace>(
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/librbd/mirror/snapshot/Utils.cc]: done.

File: src/cls/rbd/cls_rbd_types.cc (hunk 3 above)
Comment (tchaikov) [reply src/cls/rbd/cls_rbd_types.cc]: lemme try.

File: src/cls/rbd/cls_rbd_types.cc [hunk 5]
```diff
@@ -925,13 +924,13 @@ void SnapshotNamespace::decode(bufferlist::const_iterator &p)
...
-  boost::apply_visitor(DecodeSnapshotNamespaceVisitor(p), *this);
+  visit(DecodeSnapshotNamespaceVisitor(p));
   DECODE_FINISH(p);
 }
 
 void SnapshotNamespace::dump(Formatter *f) const {
-  boost::apply_visitor(
-    DumpSnapshotNamespaceVisitor(f, "snapshot_namespace_type"), *this);
+  visit(
+    DumpSnapshotNamespaceVisitor(f, "snapshot_namespace_type"));
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/cls/rbd/cls_rbd_types.cc]: true.

File: src/test/librbd/deep_copy/test_mock_SnapshotCopyRequest.cc [hunk 6]
```diff
@@ -252,7 +252,7 @@ class TestMockDeepCopySnapshotCopyRequest : public TestMockFixture {
     }
 
     if (protect) {
-      EXPECT_TRUE(boost::get<cls::rbd::UserSnapshotNamespace>(&snap_ns) !=
+      EXPECT_TRUE(std::get_if<cls::rbd::UserSnapshotNamespace>(&snap_ns) !=
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/test/librbd/deep_copy/test_mock_SnapshotCopyRequest.cc]: done.

**Comment by tchaikov** [comment]
> > because of https://cplusplus.github.io/LWG/issue3052, libstdc++
> > only specializes for std::variant<> for std::visit(), while libc++
> > allows the derived types of std::variant<> to be "visited", hence
> > we add an adaptor for SnapshotNamespace.
> 
> So libstdc++ isn't compliant here? Because https://en.cppreference.com/w/cpp/utility/variant/visit says "... or a (possibly const-qualified) class C such that there is exactly one [std::variant](https://en.cppreference.com/w/cpp/utility/variant) specialization that is a base class of C and it is a public and unambiguous base class."

tl;dr: no, at least libstdc++11 is not. that's my understanding.

long story: if you take a closer look at https://cplusplus.github.io/LWG/lwg-defects.html#3052. which suggests that the standard behavior is an issue. and libstdc++ adheres strictly to the standard, and accepts *only* the specialization of `std::variant<>`.  while libc++ accepts both the specialization and its derived classes. this behavior is proposed in a relatively new proposal, see http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2162r2.html

and in a recent change in libstdc++, P2162r2 is implemented by https://github.com/gcc-mirror/gcc/commit/c46ecb0112e91c80ee111439e79a58a953e4479d (dated Oct 2, 2021). unfortunately, this change is not included by libstdc++-11 yet.

https://en.cppreference.com/w/cpp/utility/variant/visit puts

> or a (possibly const-qualified) class C such that there is exactly one [std::variant](https://en.cppreference.com/w/cpp/utility/variant) specialization that is a base class of C and it is a public and unambiguous base class.

i think it explains the C++17 and C++20 in a human readable way after the DR of https://cplusplus.github.io/LWG/lwg-defects.html#3052 was resolved. i cross-checked the cppreference's wording with N4860 and verified it, but haven't got a chance to check the C++17 standard.

when testing Casey's C++20 changes with clang-12 + libstdc++-11, i stumped by the same error messages quoted by https://stackoverflow.com/questions/63616709/incomplete-type-stdvariant-used-in-nested-name-specifier, which in turn pointed me to the LWG issue.

File: src/cls/rbd/cls_rbd_types.cc (hunk 3 above)
Comment (tchaikov) [reply src/cls/rbd/cls_rbd_types.cc]: yup, it compiles w/o this `operator<<()`.

**Comment by tchaikov** [comment]
jenkins test make check

File: src/tools/rbd_mirror/image_replayer/snapshot/Replayer.cc [hunk 7]
```diff
@@ -768,7 +768,7 @@ void Replayer<I>::prune_non_primary_snapshot(uint64_t snap_id) {
       snap_namespace = snap_info->snap_namespace;
       snap_name = snap_info->name;
 
-      ceph_assert(boost::get<cls::rbd::MirrorSnapshotNamespace>(
+      ceph_assert(std::get_if<cls::rbd::MirrorSnapshotNamespace>(
```
[the remark this answers is not in the document]
Comment (idryomov) [reply src/tools/rbd_mirror/image_replayer/snapshot/Replayer.cc:771]: This one is still outstanding.

**Comment by idryomov** [comment]
> and in a recent change in libstdc++, P2162r2 was implemented by [gcc-mirror/gcc@c46ecb0](https://github.com/gcc-mirror/gcc/commit/c46ecb0112e91c80ee111439e79a58a953e4479d) (dated Oct 2, 2021). unfortunately, this change was not included by libstdc++-11 yet.

I see.  Thanks for the explanation!

**Comment by tchaikov** [comment]
jenkins test make check

**Review by idryomov — APPROVED** [review]
LGTM, please squash.

**Comment by tchaikov** [comment]
squashed and repushed.

File: src/librbd/api/Group.cc [hunk 8]
```diff
@@ -45,7 +45,7 @@ snap_t get_group_snap_id(I* ictx,
   for (; it != ictx->snap_ids.end(); ++it) {
     if (it->first.first == in_snap_namespace) {
       return it->second;
-    } else if (boost::get<cls::rbd::GroupSnapshotNamespace>(&it->first.first) ==
+    } else if (std::get_if<cls::rbd::GroupSnapshotNamespace>(&it->first.first) ==
```
Comment (idryomov) [inline src/librbd/api/Group.cc:48]: `!std::holds_alternative`

File: src/librbd/deep_copy/SnapshotCopyRequest.cc [hunk 9]
```diff
@@ -257,7 +257,7 @@ void SnapshotCopyRequest<I>::send_snap_remove() {
       return;
     }
 
-    if (boost::get<cls::rbd::UserSnapshotNamespace>(&snap_namespace) ==
+    if (std::get_if<cls::rbd::UserSnapshotNamespace>(&snap_namespace) ==
```
Comment (idryomov) [inline src/librbd/deep_copy/SnapshotCopyRequest.cc:260]: `!std::holds_alternative`

File: src/librbd/deep_copy/SnapshotCopyRequest.cc [hunk 10]
```diff
@@ -347,7 +347,7 @@ void SnapshotCopyRequest<I>::send_snap_create() {
 
     if (m_snap_seqs.find(src_snap_id) == m_snap_seqs.end()) {
       // the source snapshot is not in our mapping table, ...
-      if (boost::get<cls::rbd::UserSnapshotNamespace>(&snap_namespace) !=
+      if (std::get_if<cls::rbd::UserSnapshotNamespace>(&snap_namespace) !=
```
Comment (idryomov) [inline src/librbd/deep_copy/SnapshotCopyRequest.cc:350]: `std::holds_alternative`

File: src/librbd/mirror/snapshot/UnlinkPeerRequest.cc [hunk 11]
```diff
@@ -68,9 +68,9 @@ void UnlinkPeerRequest<I>::unlink_peer() {
        snap_it != m_image_ctx->snap_info.end(); ++snap_it) {
     if (snap_it->first == m_snap_id) {
       r = 0;
-      mirror_ns = boost::get<cls::rbd::MirrorSnapshotNamespace>(
+      mirror_ns = std::get_if<cls::rbd::MirrorSnapshotNamespace>(
         &snap_it->second.snap_namespace);
-    } else if (boost::get<cls::rbd::MirrorSnapshotNamespace>(
+    } else if (std::get_if<cls::rbd::MirrorSnapshotNamespace>(
```
Comment (idryomov) [inline src/librbd/mirror/snapshot/UnlinkPeerRequest.cc:73]: `std::holds_alternative`

**Review by idryomov — CHANGES_REQUESTED** [review]
A few more `std::holds_alternative` nits.  Sorry for missing that earlier!

**Comment by tchaikov** [comment]
@idryomov all fixed and repushed.

File: src/cls/rbd/cls_rbd.cc [hunk 12]
```diff
@@ -2326,7 +2326,7 @@ int snapshot_add(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
     return -EINVAL;
   }
 
-  if (boost::get<cls::rbd::UnknownSnapshotNamespace>(
+  if (std::get_if<cls::rbd::UnknownSnapshotNamespace>(
         &snap_meta.snapshot_namespace) != nullptr) {
```
Comment (idryomov) [inline src/cls/rbd/cls_rbd.cc:2330]: `std::holds_alternative`

File: src/cls/rbd/cls_rbd.cc [hunk 13]
```diff
@@ -5726,7 +5726,7 @@ int image_snapshot_unlink_peer(cls_method_context_t hctx,
     // should remove the snapshot instead.
     auto search_lambda = [snap_id](const cls_rbd_snap& snap_meta) {
       if (snap_meta.id > snap_id &&
-          boost::get<cls::rbd::MirrorSnapshotNamespace>(
+          std::get_if<cls::rbd::MirrorSnapshotNamespace>(
```
Comment (idryomov) [inline src/cls/rbd/cls_rbd.cc:5729]: `std::holds_alternative`

**Review by idryomov — CHANGES_REQUESTED** [review]
Two more -- I always forget `src/cls` in my greps...  And feel free to squash right away, I'll pull it into my testing branch immediately.

File: src/cls/rbd/cls_rbd.cc (hunk 12 above)
Comment (tchaikov) [reply src/cls/rbd/cls_rbd.cc:2330]: fixed.

File: src/cls/rbd/cls_rbd.cc (hunk 13 above)
Comment (tchaikov) [reply src/cls/rbd/cls_rbd.cc]: fixed.

**Comment by tchaikov** [comment]
fixed and repushed.

**Review by idryomov — APPROVED** [review]

**Comment by idryomov** [comment]
shaman build failed on `ceph_test_cls_rbd` binary.  It looks like a few `boost::get` call sites are remaining in `src/test/cls_rbd/test_cls_rbd.cc`.

**Comment by tchaikov** [comment]
fixed and pushed to https://shaman.ceph.com/builds/ceph/rbd-std-variant-kefu/b65fc8d9e81d37af4a27dc785846e83ccd804e40/default/297879/

**Comment by idryomov** [comment]
jenkins test make check

**Review by idryomov — APPROVED** [review]
Hopefully all of them now, please squash!

**Comment by tchaikov** [comment]
squashed and repushed.

**Comment by idryomov** [comment]
jenkins test make check

**Comment by idryomov** [comment]
No related failures:

https://pulpito.ceph.com/dis-2022-03-12_09:11:57-rbd-wip-dis-testing-distro-default-smithi/
https://pulpito.ceph.com/dis-2022-03-12_18:17:39-rbd-wip-dis-testing-distro-default-smithi/
https://pulpito.ceph.com/dis-2022-03-13_13:07:18-rbd-wip-dis-testing-distro-default-smithi/
